src/ballet/shred/fd_deshredder.h

```diff
@@ -0,0 +1,87 @@
+#ifndef HEADER_fd_src_ballet_shred_fd_deshredder_h
+#define HEADER_fd_src_ballet_shred_fd_deshredder_h
+
+#include "../fd_ballet_base.h"
+#include "fd_shred.h"
+
+FD_PROTOTYPES_BEGIN
+
+/* fd_deshredder_t: Deserializes a vector of shreds into block entries. */
+
+struct fd_deshredder {
+  /* Vector of data shreds */
+  fd_shred_t const ** shreds;
+
+  /* Number of shreds left in buffer */
+  uint shred_cnt;
+
+  /* Data offset in next shred */
+  uint data_off;
+
+  /* Cursor to target buffer
+
+     Note that this points to the end of the data
+     that was previously deserialized. */
+  uchar * buf;
+
+  /* Free space in target target buffer */
+  ulong bufsz;
+
+  /* Cached return code */
+  long result;
+};
+typedef struct fd_deshredder fd_deshredder_t;
+
+/* fd_deshredder_init: Initializes the deshredder.
+
+   `buf` is the buffer into which dconcatenated shreds get written.
+
+   `bufsz` is the size of `buf` in bytes.
+
+   `shreds` is a contiguous vector of data shreds:
+   The shred `idx` of each shred increments by exactly by one,
+   The `slot` and `version` must be the same.
+
+   `shred_cnt` is the number of `fd_shred_t` in the `shreds` buffer.
+
+   Each shred must have passed the validation checks in `fd_shred_parse`
+   and ideally should have passed authentication checks (sig verify). */
+void
+fd_deshredder_init( fd_deshredder_t *          shredder,
+                    void *                     buf,
+                    ulong                      bufsz,
+                    fd_shred_t const ** shreds,
+                    ulong                      shred_cnt );
+
+/* fd_deshredder_next: Concatenates a batch of shreds.
+
+   Concatenates a batch of shreds provided by the caller in
+   `fd_deshredder_init`.
+
+   Note that it usually takes multiple calls to process to process all
+   provided shreds because each block can have up to 64 batches.
```

Comment:
A block can have more than 64 batches, and each batch can span an arbitrary number of shreds (but an integral number of FEC sets). The `shred` tile right now is set up to do one batch = 1 FEC set, but that's not required.